components/homme/src/theta-l_kokkos/cxx/ElementsState.cpp

```diff
@@ -338,4 +342,118 @@ void ElementsState::push_to_f90_pointers (F90Ptr& state_v, F90Ptr& state_w_i, F9
+      // The isnan checks on k and k+1 are redundant except for the last k, but
+      // it's probably the fastest way to check all interface values.
+      const auto f3 = [&] (const int k) { if (h[k] < h[k+1] || isnan(h[k]) || isnan(h[k+1])) nerr = 1; };
+      const auto tvr = Kokkos::ThreadVectorRange(team, nplev);
+      parallel_for(tvr, f1);
+      parallel_for(tvr, f2);
+      parallel_for(tvr, f3);
+    };
+    parallel_for(Kokkos::TeamThreadRange(team, NP*NP), g);
+  };
+  int nerr;
+  parallel_reduce(get_default_team_policy<ExecSpace>(nelem), check, nerr);
+  
+  return nerr == 0;
+}
+
+void check_print_abort_on_bad_elems (const std::string& label, const int tlvl,
+                                     const int error_code) {
+  const auto& s = Context::singleton().get<ElementsState>();
+  
+  // On device and, thus, efficient.
+  if (all_good_elems(s, tlvl)) return;
+
+  // Now that we know there is an error, we can do the rest inefficiently.
+  const auto& geometry = Context::singleton().get<ElementsGeometry>();
+  const int nelem = s.num_elems();
+  const int nplev = NUM_PHYSICAL_LEV, ntl = NUM_TIME_LEVELS, vecsz = VECTOR_SIZE;
+  const auto& comm = Context::singleton().get<Connectivity>().get_comm();
+
+  const auto vtheta_dp_h = Kokkos::create_mirror_view(s.m_vtheta_dp);
+  Kokkos::deep_copy(vtheta_dp_h, s.m_vtheta_dp);
+  const auto dp3d_h = Kokkos::create_mirror_view(s.m_dp3d);
+  Kokkos::deep_copy(dp3d_h, s.m_dp3d);
+  const auto phinh_i_h = Kokkos::create_mirror_view(s.m_phinh_i);
+  Kokkos::deep_copy(phinh_i_h, s.m_phinh_i);
+  const auto sphere_latlon = Kokkos::create_mirror_view(geometry.m_sphere_latlon);
+  Kokkos::deep_copy(sphere_latlon, geometry.m_sphere_latlon);
+
+  HostView<Real*****>
+    vtheta_dp(reinterpret_cast<Real*>(&vtheta_dp_h(0,0,0,0,0)), nelem, ntl, NP, NP, NUM_LEV  *vecsz),
+    dp3d     (reinterpret_cast<Real*>(&dp3d_h     (0,0,0,0,0)), nelem, ntl, NP, NP, NUM_LEV  *vecsz),
+    phinh_i  (reinterpret_cast<Real*>(&phinh_i_h  (0,0,0,0,0)), nelem, ntl, NP, NP, NUM_LEV_P*vecsz);
+
+  bool first = true;
+  FILE* fid = nullptr;
+  std::string filename;
+  for (int ie = 0; ie < nelem; ++ie) {
+    for (int gi = 0; gi < NP; ++gi)
+      for (int gj = 0; gj < NP; ++gj) {
+        int k_bad = -1;
+        bool v = true, d = true, p = true;
+        for (int k = 0; k < nplev; ++k) {
+          v = std::isnan(vtheta_dp(ie,tlvl,gi,gj,k)) || vtheta_dp(ie,tlvl,gi,gj,k) < 0;
+          d = std::isnan(dp3d(ie,tlvl,gi,gj,k)) || dp3d(ie,tlvl,gi,gj,k) < 0;
+          p = (std::isnan(phinh_i(ie,tlvl,gi,gj,k)) || std::isnan(phinh_i(ie,tlvl,gi,gj,k+1)) ||
+               phinh_i(ie,tlvl,gi,gj,k) < phinh_i(ie,tlvl,gi,gj,k+1));
+          if (v || d || p) {
+            k_bad = k;
+            break;
+          }
+        }
+        if (k_bad >= 0) {
+          if (first) {
+            filename = std::string("hommexx.errlog.") + std::to_string(comm.rank());
```

Comment:
Very minor, but if we do some unit testing with multiple comm sizes, we should have the filename include the comm size, to avoid clashes of different tests. Something like `$size.$rank`.

Reply:
Good idea. I'll add the comm size to the filename.

Reply:
Done. See below. I've edited the commit message to reflect the change.